### PR TITLE
Use SDL_bool instead an int return code in the SDL API

### PR DESCRIPTION
**Pull request by slouken**
Fixes https://github.com/libsdl-org/SDL/issues/10575

### CODE REVIEW

**Comment by icculus** [comment]
Almost 71000 lines of diff.   :O

**Comment by slouken** [comment]
> Almost 71000 lines of diff. :O

Just a bit of work. :)

**Comment by slouken** [comment]
@sezero, I have an sdl2-compat patch in progress for this. I know this is a lot, and I don't expect a full review, but when you have a few minutes can you spot check the changes and see if you see any problems?

@Kontrabant, can you review the Wayland changes and make sure they make sense?

**Comment by sezero** [comment]
Is the "Use stdbool for SDL_bool" commit temporary?

**Comment by slouken** [comment]
> Is the "Use stdbool for SDL_bool" commit temporary?

No, it helps people detect when they're misusing the new API signatures. I am working on sdl2-compat changes for this.

**Comment by sezero** [comment]
> > Is the "Use stdbool for SDL_bool" commit temporary?
> 
> No, it helps people detect when they're misusing the new API signatures.

Then you should guard against c++ compilers in SDL_stdinc.h, similar to what it was in SDL_internal.h

**Comment by sezero** [comment]
.. and maybe this compile time assert to guard against broken headers / compilers?
`SDL_COMPILE_TIME_ASSERT(SDL_bool, (SDL_bool)(-1) == 1);`

**Comment by sezero** [comment]
The patchset builds for me for i686-linux. I locally applied the following for bool (see two comments above.)

```diff
diff --git a/include/SDL3/SDL_stdinc.h b/include/SDL3/SDL_stdinc.h
index 3875d4a..950faa8 100644
--- a/include/SDL3/SDL_stdinc.h
+++ b/include/SDL3/SDL_stdinc.h
@@ -36,7 +36,9 @@
 #include <inttypes.h>
 #endif
 #include <stdarg.h>
+#ifndef __cplusplus
 #include <stdbool.h>
+#endif
 #include <stdint.h>
 #include <string.h>
 #include <wchar.h>
@@ -484,6 +486,7 @@ typedef Sint64 SDL_Time;
 /** \cond */
 #ifndef DOXYGEN_SHOULD_IGNORE_THIS
 SDL_COMPILE_TIME_ASSERT(bool, sizeof(SDL_bool) == 1);
+SDL_COMPILE_TIME_ASSERT(SDL_bool, (SDL_bool)(-1) == 1);
 SDL_COMPILE_TIME_ASSERT(uint8, sizeof(Uint8) == 1);
 SDL_COMPILE_TIME_ASSERT(sint8, sizeof(Sint8) == 1);
 SDL_COMPILE_TIME_ASSERT(uint16, sizeof(Uint16) == 2);
```

**Comment by slouken** [comment]
> The patchset builds for me for i686-linux. I locally applied the following for bool (see two comments above.)

Added, thanks!

**Comment by sezero** [comment]
> Added, thanks!

Thinking about it, maybe `SDL_COMPILE_TIME_ASSERT(negative_bool, (SDL_bool)(-1) == 1);`
is overzealous? A compiler might not have `_Bool` and define bool as `unsigned char`
and the assert above would fail in that case. Maybe something like the following so as
to provide user a bit freedom? (`HAVE_BROKEN__BOOL` being a user macro)

```diff
diff --git a/include/SDL3/SDL_stdinc.h b/include/SDL3/SDL_stdinc.h
index 133c2f8..439d020 100644
--- a/include/SDL3/SDL_stdinc.h
+++ b/include/SDL3/SDL_stdinc.h
@@ -488,3 +488,6 @@ typedef Sint64 SDL_Time;
 SDL_COMPILE_TIME_ASSERT(bool, sizeof(SDL_bool) == 1);
+#if defined(__cplusplus) || !defined(HAVE_BROKEN__BOOL)
 SDL_COMPILE_TIME_ASSERT(negative_bool, (SDL_bool)(-1) == 1);
+#endif
+SDL_COMPILE_TIME_ASSERT(negative_bool2, (SDL_bool)(-1) > 0);
 SDL_COMPILE_TIME_ASSERT(uint8, sizeof(Uint8) == 1);
```

(And maybe I'm being more overzealous now...)

File: docs/README-migration.md [hunk 1]
```diff
@@ -1664,6 +1696,8 @@ This header has been removed and a simplified version of this API has been added
 The standard C headers like stdio.h and stdlib.h are no longer included, you should include them directly in your project if you use non-SDL C runtime functions.
 M_PI is no longer defined in SDL_stdinc.h, you can use the new symbols SDL_PI_D (double) and SDL_PI_F (float) instead.
 
+SDL_bool is now defined as bool, and is 1 byte instead of the size of an int.
```
Comment (sezero) [inline docs/README-migration.md:1699]: Maybe add here something like `Presence of a conformant stdbool.h is required for C code.` ??

Comment (slouken) [reply docs/README-migration.md:1699]: I don't think we need that. It'll be pretty obvious if SDL fails to compile, missing stdbool.h, and if it doesn't, people don't need to think about it.

Comment (sezero) [reply docs/README-migration.md:1699]: OK

**Comment by Kontrabant** [comment]
> @Kontrabant, can you review the Wayland changes and make sure they make sense?

The changes look good to me. Definitely an impressively sized diff :)

Some pending pulls waiting on new upstream library versions to be tagged, like the Wayland icon protocol support, will need to be updated, and it looks like there are a couple of places where static internal functions returning an int can become booleans for cleaner pass-through of return values, but that can come later.

**Comment by sezero** [comment]
All satellite libs also need updating after this.

**Comment by slouken** [comment]
> All satellite libs also need updating after this.

Already done. I haven't changed their APIs to this model yet, but they build and work with the latest SDL.

**Comment by sezero** [comment]
> > All satellite libs also need updating after this.
> 
> Already done. I haven't changed their APIs to this model yet, but they build and work with the latest SDL.

SDL_net and SDL_rtf still need updating

**Comment by slouken** [comment]
> > > All satellite libs also need updating after this.
> > 
> > 
> > Already done. I haven't changed their APIs to this model yet, but they build and work with the latest SDL.
> 
> SDL_net and SDL_rtf still need updating

Done!

**Comment by sezero** [comment]
> 
> Done!

Pushed further updates to SDL_image and SDL_mixer for SDL_SetError now returning SDL_FALSE, not -1

**Comment by slouken** [comment]
> > Done!
> 
> Pushed further updates to SDL_image and SDL_mixer for SDL_SetError now returning SDL_FALSE, not -1

Thanks. I'm going through and updating those libraries more carefully now.

**Comment by sezero** [comment]
BTW, wiki is failing: https://github.com/libsdl-org/sdlwiki/actions/runs/10585301989/job/29331580831

**Comment by slouken** [comment]
> BTW, wiki is failing: https://github.com/libsdl-org/sdlwiki/actions/runs/10585301989/job/29331580831

Thanks, I'll take a look.

**Comment by slouken** [comment]
> > BTW, wiki is failing: https://github.com/libsdl-org/sdlwiki/actions/runs/10585301989/job/29331580831
> 
> Thanks, I'll take a look.

Fixed, thanks!

**Comment by Sackzement** [comment]
`bool` is a C99 feature. Does this change mean, that SDL headers now require the  C99 standard?

Maybe worth mentioning:
Previously, when casting a floating-point type to `int`, the values  between -1.0 < val < +1.0 would round to 0 and therefore be `false`.
Now, when casting a floating-point type to `bool`, all values are `true`, only the value 0.0 is `false`.

**Comment by slouken** [comment]
> `bool` is a C99 feature. Does this change mean, that SDL headers now require the C99 standard?

No, but they do require a stdbool.h that defines bool as a 1 byte type. You can create one yourself if your build environment doesn't provide it.

> Maybe worth mentioning: Previously, when casting a floating-point type to `int`, the values between -1.0 < val < +1.0 would round to 0 and therefore be `false`. Now, when casting a floating-point type to `bool`, all values are `true`, only the value 0.0 is `false`.

Yep, hopefully nobody was casting a float value to an SDL return code.

**Pull request merged**